django/db/models/fields/composite.py

```diff
@@ -0,0 +1,192 @@
+from django.db.models.expressions import Col, ExpressionList
+from django.db.models.fields import Field
+from django.db.models.lookups import TupleExact
+from django.db.models.query_utils import DeferredAttribute
+from django.db.models.signals import class_prepared
+from django.utils.functional import cached_property
+
+
+class CompositeCol(ExpressionList):
+    def __init__(self, alias, target: "CompositeField", output_field=None):
+        if output_field is None:
+            output_field = target
+        self.cols = [Col(alias, field) for field in target.component_fields]
+        super().__init__(*self.cols, output_field=output_field)
+        self.alias, self.target = alias, target
+
+    def as_sql(self, compiler, connection):
+        sql, params = super().as_sql(compiler, connection)
+        return "(%s)" % sql, params
+
+    def get_lookup(self, lookup):
+        if lookup == "exact":
+            return TupleExact
```

Comment:
👋 @LilyFoote! Just passing by as [there was some recent interest in this area from the forum](https://forum.djangoproject.com/t/multiple-columns-using-mysql-in-clause/24488/7) to mention that if you're looking to implement `TupleIn` there is some pre-existing logic to support it for joins at least.

See ticket-19385, and the `ForeignObject`, `MultiColSource`, `SQLQuery.build_filter`, `RelatedIn` symbols for reference.

Reply:
Thanks for the pointers! Where I'm currently stuck is working out how to adapt `ForeignKey` to point to a `CompositeField`. There's something about how `ForeignKey` maps to the database column(s) that's just not clicked yet.